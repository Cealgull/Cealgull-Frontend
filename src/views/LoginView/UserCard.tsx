import { useNavigation } from "@react-navigation/native";
import { Avatar, Icon, Image } from "@rneui/themed";
import { LoginTabScreenPropsGeneric } from "@src/@types/navigation";
import React, { useMemo } from "react";
import {
  Dimensions,
  Pressable,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewStyle,
} from "react-native";

interface UserCardProps {
  userName: string;
  email: string;
  selected: boolean;
}

export default function UserCard({ userName, email, selected }: UserCardProps) {
  const checkIconStyle = useMemo<StyleProp<ViewStyle>>(
    () => [styles.check_container, { opacity: selected ? 100 : 0 }],
    [selected]
  );

  return (
    <View style={styles.card}>
      <View style={styles.content_container}>
        <View style={{ flex: 0 }}>
          <UserCardAvatar alt={userName} />
        </View>
        <View style={styles.info_container}>
          <Text style={[styles.text, styles.text_username]}>{userName}</Text>
          <Text style={styles.text}>{email}</Text>
        </View>
      </View>
      <View style={checkIconStyle} testID="check_icon">
        <CheckIcon />
      </View>
    </View>
  );
}

/**
 * Placeholder for no user.
 */
export function UserAddCard() {
  const navigation =
    useNavigation<LoginTabScreenPropsGeneric<"UserLogin">["navigation"]>();

  return (
    <Pressable
      onPress={() => {
        navigation.navigate("UserAdd");
      }}
    >
      <View style={styles.card}>
        <View style={styles.content_container}>
          <View style={{ flex: 0 }}>
            <Icon name="adduser" type="antdesign" size={72} />
          </View>
          <View style={styles.info_container}>
            <Text style={{ fontSize: 20 }}>添加用户</Text>
          </View>
        </View>
      </View>
    </Pressable>
  );
}

/**
 * The green check mark icon
 */
const checkMarkIconPath = "../../../assets/check-mark-icon.png";
const CheckIcon: React.FC = () => {
  return (
    <Image source={require(checkMarkIconPath)} style={styles.check_icon} />
  );
};

interface UserCardAvatarProps {
  size?: ("small" | "medium" | "large" | "xlarge") | number;
  uri?: string;
  alt?: string;
}

const UserCardAvatar: React.FC<UserCardAvatarProps> = ({
  size = 80,
  uri = "https://randomuser.me/api/portraits/men/36.jpg",
  alt,
}) => {
  return <Avatar rounded size={size} source={{ uri }} imageProps={{ alt }} />;
};

const { height: screenHeight, width: screenWidth } = Dimensions.get("screen");

const styles = StyleSheet.create({
  card: {
    backgroundColor: "white",
    borderColor: "rgba(0,0,0,0.7)",
    borderRadius: 10,
    borderWidth: 3,
    height: screenHeight / 8,
    margin: 16,
    shadowColor: "black",
    shadowOffset: { width: 8, height: 8 },
    shadowOpacity: 0.1,
    shadowRadius: 6,
    width: screenWidth - 80,
  },
  check_container: {
    backgroundColor: "transparent",
    position: "absolute",
    right: 0,
    top: 0,
    zIndex: 999,
  },
  check_icon: {
    ...(function () {
      const x = screenHeight / 8;
      return { height: x, width: x };
    })(),
  },
  content_container: {
    alignItems: "center",
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    marginHorizontal: 10,
  },
  info_container: {
    alignItems: "center",
    flex: 1,
    justifyContent: "space-between",
  },
  text: {
    marginVertical: 6,
  },
  text_username: {
    fontSize: 20,
    fontWeight: "700",
  },
});
