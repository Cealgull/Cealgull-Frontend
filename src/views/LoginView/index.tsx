/**
 * @author Bojun Ren
 * @data 2023/07/07
 */
export { default } from "./LoginView";
