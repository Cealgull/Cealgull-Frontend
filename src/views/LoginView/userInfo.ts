// TODO real information
interface UserInfo {
  userName: string;
  email: string;
}

const userInfoList: UserInfo[] = [
  { userName: "Dean Chambers", email: "tih@fe.gi" },
];

export default userInfoList;
