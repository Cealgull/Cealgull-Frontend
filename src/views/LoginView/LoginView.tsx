import { useNavigation } from "@react-navigation/native";
import { Button, ButtonProps } from "@rneui/themed";
import {
  LoginTabScreenPropsGeneric,
  StackScreenPropsGeneric,
} from "@src/@types/navigation";
import { Loadable } from "@src/components/Loadable";
import config from "@src/config";
import { User } from "@src/models/User";
import React, { useCallback, useEffect, useState } from "react";
import { Dimensions, Pressable, StyleSheet, Text, View } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import UserCard, { UserAddCard } from "./UserCard";
import { useSetUser } from "@src/hooks/useUser";

const userLengthMax = config["login.user.max"];

export default function LoginView() {
  const [selected, setSelected] = useState<number | undefined>(undefined);
  const rootNavigation =
    useNavigation<StackScreenPropsGeneric<"Login">["navigation"]>();
  const loginNavigation =
    useNavigation<LoginTabScreenPropsGeneric<"UserLogin">["navigation"]>();
  const [userList, setUserList] = useState<User[] | undefined>(undefined);
  const [disableLogin, setDisableLogin] = useState(false);
  const setUser = useSetUser();

  const handleLogin = async () => {
    setDisableLogin(true);
    const user = (userList as User[])[selected as number];
    try {
      await user.login();
      await user.persist();
      setUser(user);
      rootNavigation.navigate("Main");
    } catch (err) {
      console.error(err);
    } finally {
      setDisableLogin(false);
    }
  };
  const handleDeleteUser = async () => {
    await (userList as User[])[selected as number].detach();
    // After one user is detached, the other users' id are changed.
    // Therefore, we reset the state `userList`.
    setUserList(undefined);
    setSelected(undefined);
    setUserList(await createUserList());
  };
  const handleAddUser = useCallback(() => {
    loginNavigation.navigate("UserAdd");
  }, [loginNavigation]);

  async function createUserList() {
    const userCount = await User.getUserCount();
    const res: Array<User> = [];
    for (let i = 0; i < userCount; ++i) {
      // Get the user from local storage
      const user = await User.getUser(i);
      // Examine if the user's profile is persisted
      user.hasProfile() ? res.push(user) : user.detach();
    }
    return res;
  }
  useEffect(() => {
    createUserList().then((res) => {
      setUserList(res);
    });
  }, []);

  const createUserCardList = useCallback(
    (userList: User[], selected?: number) => {
      const userCardList: React.ReactElement[] = [];
      userList.forEach((user, i) => {
        const { username, signature, avatar } = user.profile;
        userCardList.push(
          <UserCard
            key={`user_login${i}`}
            username={username}
            signature={signature}
            avatar={avatar || undefined}
            selected={selected === i}
            onPress={() => setSelected(i)}
          />
        );
      });

      for (let i = userList.length; i < userLengthMax; ++i) {
        userCardList.push(
          <Pressable
            key={`user_login${i}`}
            onPress={handleAddUser}
            accessibilityRole="button"
          >
            <UserAddCard />
          </Pressable>
        );
      }
      return userCardList;
    },
    [handleAddUser]
  );

  return (
    <SafeAreaView style={styles.page}>
      {/* A '！' (U+FF01) is ugly, since it make the text bias */}
      <Text style={styles.welcome}>欢迎回来</Text>
      <View style={styles.user_list}>
        <Loadable
          generator={(_userList) => createUserCardList(_userList, selected)}
          maybeProp={userList}
        />
      </View>
      <View>
        <LoginButton
          onPress={handleLogin}
          disabled={disableLogin || selected === undefined}
        />
        <DelUserButton
          onPress={handleDeleteUser}
          disabled={selected === undefined}
        />
      </View>
    </SafeAreaView>
  );
}

type LoginButtonProps = Omit<
  ButtonProps,
  "title" | "buttonStyle" | "titleStyle" | "containerStyle"
>;

const LoginButton: React.FC<LoginButtonProps> = (props) => {
  return (
    <Button
      title="登入"
      buttonStyle={[styles.button, styles.button_login]}
      titleStyle={styles.button_text}
      containerStyle={styles.button_container}
      {...props}
    />
  );
};

type DelUserButtonProps = LoginButtonProps;
const DelUserButton: React.FC<DelUserButtonProps> = (props) => {
  return (
    <Button
      title={"删除用户"}
      buttonStyle={[styles.button, styles.button_add_user]}
      titleStyle={styles.button_text}
      containerStyle={styles.button_container}
      {...props}
    />
  );
};

const { height: screenHeight } = Dimensions.get("screen");

const styles = StyleSheet.create({
  button: {
    borderRadius: 5,
  },
  button_login: {
    backgroundColor: "#48b89f",
  },
  button_add_user: {
    backgroundColor: "#d62828",
  },
  button_container: {
    height: 50,
    marginHorizontal: 50,
    marginVertical: 10,
    width: 200,
  },
  button_text: {
    fontSize: 23,
    fontWeight: "bold",
  },
  page: {
    alignItems: "center",
    flex: 1,
    justifyContent: "space-evenly",
  },
  welcome: {
    fontSize: 30,
    fontWeight: "700",
  },
  user_list: {
    height: (screenHeight / 8) * 3.6,
    justifyContent: "center",
  },
});
