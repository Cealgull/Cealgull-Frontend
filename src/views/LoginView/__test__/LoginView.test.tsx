/**
 * @author Bojun Ren
 * @data 2023/07/07
 */
import { fireEvent, render, screen } from "@testing-library/react-native";
import { ReactTestInstance } from "react-test-renderer";
import LoginView from "../LoginView";

// FIXME use mock data
test("render LoginView correctly", () => {
  render(<LoginView />);
  const userCards = screen.getAllByRole("checkbox");
  expect(userCards.length).toBeGreaterThanOrEqual(1);
});

test("select user correctly", () => {
  function testOpacityToBe(x: number, value: number) {
    const checkIcon = screen.getAllByTestId("check_icon").at(x);
    expect(checkIcon?.props.style.at(-1).opacity).toBe(value);
  }

  render(<LoginView />);

  const userCards = screen.getAllByRole("checkbox");
  testOpacityToBe(0, 0);
  testOpacityToBe(1, 0);
  testOpacityToBe(2, 0);

  fireEvent.press(userCards.at(0) as ReactTestInstance);

  testOpacityToBe(0, 100);
  testOpacityToBe(1, 0);
  testOpacityToBe(2, 0);

  fireEvent.press(userCards.at(1) as ReactTestInstance);

  testOpacityToBe(0, 0);
  testOpacityToBe(1, 100);
  testOpacityToBe(2, 0);
});
