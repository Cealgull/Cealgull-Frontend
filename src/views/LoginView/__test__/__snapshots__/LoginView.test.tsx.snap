// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LoginView render correctly 1`] = `
<View
  style={
    {
      "bottom": 34,
      "left": 0,
      "right": 0,
      "top": 47,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 30,
        "fontWeight": "700",
      }
    }
  >
    欢迎回来
  </Text>
  <View
    style={
      {
        "height": 600.3000000000001,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "white",
          "borderColor": "rgba(0,0,0,0.7)",
          "borderRadius": 10,
          "borderWidth": 3,
          "height": 166.75,
          "margin": 16,
          "shadowColor": "black",
          "shadowOffset": {
            "height": 8,
            "width": 8,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 6,
          "width": 670,
        }
      }
    >
      <View
        accessibilityRole="checkbox"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginHorizontal": 10,
            }
          }
        >
          <View
            style={
              {
                "flex": 0,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "transparent",
                  "borderRadius": 40,
                  "height": 80,
                  "width": 80,
                }
              }
              theme={
                {
                  "colors": {
                    "background": "#ffffff",
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ad1457",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                  "spacing": {
                    "lg": 12,
                    "md": 8,
                    "sm": 4,
                    "xl": 24,
                    "xs": 2,
                  },
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 40,
                    "flex": 1,
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                <Image
                  alt="john_doe"
                  borderRadius={40}
                  onLoad={[Function]}
                  source={
                    {
                      "uri": "https://randomuser.me/api/portraits/men/36.jpg",
                    }
                  }
                  style={
                    {
                      "bottom": 0,
                      "flex": 1,
                      "height": undefined,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "width": undefined,
                    }
                  }
                  testID="RNE__Avatar__Image"
                  transitionDuration={360}
                />
                <View
                  accessibilityElementsHidden={true}
                  collapsable={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    {
                      "bottom": 0,
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "#bdbdbd",
                        "flex": 1,
                        "height": undefined,
                        "justifyContent": "center",
                        "width": undefined,
                      }
                    }
                    testID="RNE__Image__placeholder"
                  />
                </View>
                <View
                  style={
                    {
                      "flex": 1,
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  testID="RNE__Image__children__container"
                />
              </View>
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "space-between",
              }
            }
          >
            <Text
              style={
                [
                  {
                    "marginVertical": 6,
                  },
                  {
                    "fontSize": 20,
                    "fontWeight": "700",
                  },
                ]
              }
            >
              john_doe
            </Text>
            <Text
              style={
                {
                  "marginVertical": 6,
                }
              }
            >
              Hello, world!
            </Text>
          </View>
        </View>
        <View
          style={
            [
              {
                "backgroundColor": "transparent",
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 999,
              },
              {
                "opacity": 0,
              },
            ]
          }
          testID="check_icon"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              {
                "backgroundColor": "transparent",
                "overflow": "hidden",
                "position": "relative",
              }
            }
          >
            <Image
              onLoad={[Function]}
              source={1}
              style={
                {
                  "bottom": 0,
                  "height": 166.75,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 166.75,
                }
              }
              testID="RNE__Image"
              theme={
                {
                  "colors": {
                    "background": "#ffffff",
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ad1457",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                  "spacing": {
                    "lg": 12,
                    "md": 8,
                    "sm": 4,
                    "xl": 24,
                    "xs": 2,
                  },
                }
              }
              transitionDuration={360}
            />
            <View
              accessibilityElementsHidden={true}
              collapsable={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#bdbdbd",
                    "height": 166.75,
                    "justifyContent": "center",
                    "width": 166.75,
                  }
                }
                testID="RNE__Image__placeholder"
              />
            </View>
            <View
              style={
                {
                  "height": 166.75,
                  "width": 166.75,
                }
              }
              testID="RNE__Image__children__container"
            />
          </View>
        </View>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          {
            "backgroundColor": "white",
            "borderColor": "rgba(0,0,0,0.7)",
            "borderRadius": 10,
            "borderWidth": 3,
            "height": 166.75,
            "margin": 16,
            "shadowColor": "black",
            "shadowOffset": {
              "height": 8,
              "width": 8,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 6,
            "width": 670,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginHorizontal": 10,
              }
            }
          >
            <View
              style={
                {
                  "flex": 0,
                }
              }
            >
              <View
                style={
                  {
                    "overflow": "hidden",
                  }
                }
                testID="RNE__ICON__CONTAINER"
              >
                <View
                  accessibilityRole="button"
                  android_ripple={
                    {
                      "borderless": false,
                      "color": "rgba(0, 0, 0, 0.3)",
                      "radius": -5,
                    }
                  }
                  disabled={false}
                  testID="RNE__ICON__CONTAINER_ACTION"
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    testID="RNE__ICON"
                  >
                    <Text
                      allowFontScaling={false}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "#242424",
                            "fontSize": 72,
                          },
                          {
                            "backgroundColor": "transparent",
                          },
                          {
                            "fontFamily": "anticon",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          {},
                        ]
                      }
                      testID="RNE__ICON__Component"
                    >
                      
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  {
                    "fontSize": 20,
                  }
                }
              >
                添加用户
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          {
            "backgroundColor": "white",
            "borderColor": "rgba(0,0,0,0.7)",
            "borderRadius": 10,
            "borderWidth": 3,
            "height": 166.75,
            "margin": 16,
            "shadowColor": "black",
            "shadowOffset": {
              "height": 8,
              "width": 8,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 6,
            "width": 670,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginHorizontal": 10,
              }
            }
          >
            <View
              style={
                {
                  "flex": 0,
                }
              }
            >
              <View
                style={
                  {
                    "overflow": "hidden",
                  }
                }
                testID="RNE__ICON__CONTAINER"
              >
                <View
                  accessibilityRole="button"
                  android_ripple={
                    {
                      "borderless": false,
                      "color": "rgba(0, 0, 0, 0.3)",
                      "radius": -5,
                    }
                  }
                  disabled={false}
                  testID="RNE__ICON__CONTAINER_ACTION"
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    testID="RNE__ICON"
                  >
                    <Text
                      allowFontScaling={false}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "#242424",
                            "fontSize": 72,
                          },
                          {
                            "backgroundColor": "transparent",
                          },
                          {
                            "fontFamily": "anticon",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          {},
                        ]
                      }
                      testID="RNE__ICON__Component"
                    >
                      
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  {
                    "fontSize": 20,
                  }
                }
              >
                添加用户
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          {
            "height": 50,
            "marginHorizontal": 50,
            "marginVertical": 10,
            "width": 200,
          },
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "hsl(208, 8%, 90%)",
              "borderColor": "#2089dc",
              "borderRadius": 5,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "hsl(208, 8%, 63%)",
                "fontSize": 23,
                "fontWeight": "bold",
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            登入
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          {
            "height": 50,
            "marginHorizontal": 50,
            "marginVertical": 10,
            "width": 200,
          },
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "hsl(208, 8%, 90%)",
              "borderColor": "#2089dc",
              "borderRadius": 5,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "hsl(208, 8%, 63%)",
                "fontSize": 23,
                "fontWeight": "bold",
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            删除用户
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
